README.md
<div>
<img align="left" src="../assets/peerwatch.ico?raw=true" alt="PeerWatch" width="50px" height="50px" />

# PeerWatch &nbsp; ![GitHub release (latest by date)](https://img.shields.io/github/v/release/cipheras/peerwatch?style=flat-square&logo=superuser) 
  
</div>
  
#### A tool to stream videos directly into the media player without any hazzle by just name of the video. It also have UI version to make it easy for common users. To run UI version, just execute binary directly.

![GO version](https://img.shields.io/github/go-mod/go-version/cipheras/peerwatch?style=flat-square&color=blue)
&nbsp;&nbsp;&nbsp;&nbsp;
![GitHub All Releases](https://img.shields.io/github/downloads/cipheras/peerwatch/total?style=flat-square&color=darkgreen)
&nbsp;&nbsp;&nbsp;&nbsp;
![platform](https://img.shields.io/badge/dynamic/json?url=https://jsonkeeper.com/b/KNO7&label=platform&query=platform&style=flat-square&labelColor=grey&color=purple)

![gif](../assets/screen.gif?raw=true)


## Download
Download the tool from here:
Windows | Linux
--------|-------
[win-x64](https://github.com/cipheras/peerwatch/releases/download/v2.2/peerwatch-x64.zip) | [linux-x64](https://github.com/cipheras/peerwatch/releases/download/v2.2/peerwatch-x64)


## Building
You can use tool's pre-compiled binaries directly or you can compile from source.<br>
To build from source, GO must be installed.<br><br>
For linux installation:
```
sudo make
```
For linux build:
```
sudo make build
```
For linux uninstall:
```
sudo make uninstall
```
For windows:<br>
*NOTE: For windows, DLLs are necessary to be kept in the same directory.*
```
go build
```


## Usage
**It may take some time to buffer as it depends on the file size and network speed.**
```
[+] ./peerwatch [name] or ./peerwatch "[name with multiple words]"
[+] Press ctrl + c to close

  -p int
        Port on which the tool will stream (default 8080)
  -player string
        Video player to play video [vlc | mpv | mplayer] (default "vlc")
  -tcp
        Connection over TCP or UDP (default true)
  -tp int
        Port on which the engine will work (default 50010)

Example:
peerwatch.exe xyz
or
./peerwatch "xyz blueray"
```

## Disclaimer
*This tool or author are not responsible for any type of copywrite claim. This tool work similarly as common user visiting the source and streaming videos on their browsers.*

## License
**peerwatch** is made by **@cipheras** and is released under the terms of the &nbsp;![GitHub License](https://img.shields.io/github/license/cipheras/peerwatch?color=darkgreen)


## Contact &nbsp; [![Twitter](https://img.shields.io/twitter/url?style=social&url=https%3A%2F%2Fgithub.com%2Fcipheras%2Fpeerwatch&label=Tweet)](https://twitter.com/intent/tweet?text=Hi:&url=https%3A%2F%2Fgithub.com%2Fcipheras%2Fpeerwatch)
> Feel free to submit a bug, add features or issue a pull request.



CHANGELOG
v2.2
- Added multiple server support
- Status and speed info in UI


v2.1
- Added new compiled working DLLs for windows
- Resolved some goroutine issues
- Linked window close with program termination
- Added support for in-window progress status(UI changes are left) 


v2.0
- UI support
- Multiple players to choose from, to play videos [vlc | mpv | mplayer]
- Auto update check feature
- Port to choose for connections and streaming seperately
- Choice over TCP or UDP connections 
- More optimized code


src/client.go
package src

import (
	"bufio"
	"compress/gzip"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"regexp"
	"strings"
	"time"

	"github.com/anacrolix/torrent"
	"github.com/anacrolix/torrent/iplist"
	"github.com/dustin/go-humanize"
)

const clearScreen = "\033[H\033[2J"

const torrentBlockListURL = "http://john.bitsurge.net/public/biglist.p2p.gz"

var isHTTP = regexp.MustCompile(`^https?:\/\/`)

var DS string

// ClientError formats errors coming from the client.
type ClientError struct {
	Type   string
	Origin error
}

func (clientError ClientError) Error() string {
	return fmt.Sprintf("Error %s: %s\n", clientError.Type, clientError.Origin)
}

// Client manages the torrent downloading.
type Client struct {
	Client   *torrent.Client
	Torrent  *torrent.Torrent
	Progress int64
	Uploaded int64
	Config   ClientConfig
}

// ClientConfig specifies the behaviour of a client.
type ClientConfig struct {
	TorrentPath    string
	Port           int
	TorrentPort    int
	Seed           bool
	TCP            bool
	MaxConnections int
}

// // NewClientConfig creates a new default configuration.
// func NewClientConfig() ClientConfig {
// 	return ClientConfig{
// 		Port:           8080,
// 		TorrentPort:    50007,
// 		Seed:           false,
// 		TCP:            true,
// 		MaxConnections: 200,
// 	}
// }

// NewClient creates a new torrent client based on a magnet or a torrent file.
// If the torrent file is on http, we try downloading it.
func NewClient(cfg ClientConfig) (client Client, err error) {
	var t *torrent.Torrent
	var c *torrent.Client

	client.Config = cfg

	blocklist := getBlocklist()
	torrentConfig := torrent.NewDefaultClientConfig()
	torrentConfig.DataDir = os.TempDir()
	torrentConfig.NoUpload = !cfg.Seed
	torrentConfig.DisableTCP = !cfg.TCP
	torrentConfig.ListenPort = cfg.TorrentPort
	torrentConfig.IPBlocklist = blocklist

	// Create client.
	c, err = torrent.NewClient(torrentConfig)

	if err != nil {
		return client, ClientError{Type: "creating torrent client", Origin: err}
	}

	client.Client = c

	// Add torrent.

	// Add as magnet url.
	if strings.HasPrefix(cfg.TorrentPath, "magnet:") {
		if t, err = c.AddMagnet(cfg.TorrentPath); err != nil {
			return client, ClientError{Type: "adding torrent", Origin: err}
		}
	} else {
		// Otherwise add as a torrent file.

		// If it's online, we try downloading the file.
		if isHTTP.MatchString(cfg.TorrentPath) {
			if cfg.TorrentPath, err = downloadFile(cfg.TorrentPath); err != nil {
				return client, ClientError{Type: "downloading torrent file", Origin: err}
			}
		}

		if t, err = c.AddTorrentFromFile(cfg.TorrentPath); err != nil {
			return client, ClientError{Type: "adding torrent to the client", Origin: err}
		}
	}

	client.Torrent = t
	client.Torrent.SetMaxEstablishedConns(cfg.MaxConnections)

	go func() {
		<-t.GotInfo()
		t.DownloadAll()

		// Prioritize first 5% of the file.
		largestFile := client.getLargestFile()
		firstPieceIndex := largestFile.Offset() * int64(t.NumPieces()) / t.Length()
		endPieceIndex := (largestFile.Offset() + largestFile.Length()) * int64(t.NumPieces()) / t.Length()
		for idx := firstPieceIndex; idx <= endPieceIndex*5/100; idx++ {
			t.Piece(int(idx)).SetPriority(torrent.PiecePriorityNow)
		}
	}()

	return
}

// Download and add the blocklist.
func getBlocklist() iplist.Ranger {
	var err error
	blocklistPath := os.TempDir() + "/go-peerflix-blocklist.gz"

	if _, err = os.Stat(blocklistPath); os.IsNotExist(err) {
		err = downloadBlockList(blocklistPath)
	}

	if err != nil {
		log.Printf("Error downloading blocklist: %s", err)
		return nil
	}

	// Load blocklist.
	// #nosec
	// We trust our temporary directory as we just wrote the file there ourselves.
	blocklistReader, err := os.Open(blocklistPath)
	if err != nil {
		log.Printf("Error opening blocklist: %s", err)
		return nil
	}

	// Extract file.
	gzipReader, err := gzip.NewReader(blocklistReader)
	if err != nil {
		log.Printf("Error extracting blocklist: %s", err)
		return nil
	}

	// Read as iplist.
	blocklist, err := iplist.NewFromReader(gzipReader)
	if err != nil {
		log.Printf("Error reading blocklist: %s", err)
		return nil
	}

	log.Printf("Loading blocklist.\nFound %d ranges\n", blocklist.NumRanges())
	return blocklist
}

func downloadBlockList(blocklistPath string) (err error) {
	log.Printf("Downloading blocklist")
	fileName, err := downloadFile(torrentBlockListURL)
	if err != nil {
		log.Printf("Error downloading blocklist: %s\n", err)
		return
	}

	return os.Rename(fileName, blocklistPath)
}

// Close cleans up the connections.
func (c *Client) Close() {
	c.Torrent.Drop()
	c.Client.Close()
}

// Render outputs the command line interface for the client.
func (c *Client) Render() {
	t := c.Torrent

	if t.Info() == nil {
		return
	}

	currentProgress := t.BytesCompleted()
	downloadSpeed := humanize.Bytes(uint64(currentProgress-c.Progress)) + "/s"
	c.Progress = currentProgress

	complete := humanize.Bytes(uint64(currentProgress))
	size := humanize.Bytes(uint64(t.Info().TotalLength()))

	bytesWrittenData := t.Stats().BytesWrittenData
	uploadProgress := (&bytesWrittenData).Int64() - c.Uploaded
	uploadSpeed := humanize.Bytes(uint64(uploadProgress)) + "/s"
	c.Uploaded = uploadProgress
	percentage := c.percentage()
	totalLength := t.Info().TotalLength()

	output := bufio.NewWriter(os.Stdout)

	fmt.Fprint(output, clearScreen)
	fmt.Fprint(output, t.Info().Name+"\n")
	fmt.Fprint(output, strings.Repeat("=", len(t.Info().Name))+"\n")
	if c.ReadyForPlayback() {
		fmt.Fprintf(output, "Streaming at: \thttp://localhost:%d\n", c.Config.Port)
	} else {
		// If not ready for stream
		fmt.Fprint(output, ">> Stream will start after 5% progress <<\n\n")
	}
	if currentProgress > 0 {
		fmt.Fprintf(output, "Progress: \t%s / %s  %.2f%%\n", complete, size, percentage)
	}
	if currentProgress < totalLength {
		DS = downloadSpeed // for GUI
		fmt.Fprintf(output, "Download speed: %s\n", downloadSpeed)
	}
	if c.Config.Seed {
		fmt.Fprintf(output, "Upload speed: \t%s", uploadSpeed)
	}

	output.Flush()
}

func (c Client) getLargestFile() *torrent.File {
	var target *torrent.File
	var maxSize int64

	for _, file := range c.Torrent.Files() {
		if maxSize < file.Length() {
			maxSize = file.Length()
			target = file
		}
	}

	return target
}

/*
func (c Client) RenderPieces() (output string) {
	pieces := c.Torrent.PieceStateRuns()
	for i := range pieces {
		piece := pieces[i]

		if piece.Priority == torrent.PiecePriorityReadahead {
			output += "!"
		}

		if piece.Partial {
			output += "P"
		} else if piece.Checking {
			output += "c"
		} else if piece.Complete {
			output += "d"
		} else {
			output += "_"
		}
	}

	return
}
*/

// ReadyForPlayback checks if the torrent is ready for playback or not.
// We wait until 5% of the torrent to start playing.
func (c Client) ReadyForPlayback() bool {
	return c.percentage() > 5
}

// GetFile is an http handler to serve the biggest file managed by the client.
func (c Client) GetFile(w http.ResponseWriter, r *http.Request) {
	target := c.getLargestFile()
	entry, err := NewFileReader(target)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	defer func() {
		if err := entry.Close(); err != nil {
			log.Printf("Error closing file reader: %s\n", err)
		}
	}()

	w.Header().Set("Content-Disposition", "attachment; filename=\""+c.Torrent.Info().Name+"\"")
	http.ServeContent(w, r, target.DisplayPath(), time.Now(), entry)
}

func (c Client) percentage() float64 {
	info := c.Torrent.Info()

	if info == nil {
		return 0
	}

	return float64(c.Torrent.BytesCompleted()) / float64(info.TotalLength()) * 100
}

func downloadFile(URL string) (fileName string, err error) {
	var file *os.File
	if file, err = ioutil.TempFile(os.TempDir(), "go-peerflix"); err != nil {
		return
	}

	defer func() {
		if ferr := file.Close(); ferr != nil {
			log.Printf("Error closing torrent file: %s", ferr)
		}
	}()

	// #nosec
	// We are downloading the url the user passed to us, we trust it is a torrent file.
	response, err := http.Get(URL)
	if err != nil {
		return
	}

	defer func() {
		if ferr := response.Body.Close(); ferr != nil {
			log.Printf("Error closing torrent file: %s", ferr)
		}
	}()

	_, err = io.Copy(file, response.Body)

	return file.Name(), err
}


src/player.go
package src

import (
	"log"
	"os/exec"
	"runtime"
	"strconv"
	"strings"
)

// GenericPlayer represents most players. The stream URL will be appended to the arguments.
type GenericPlayer struct {
	Name string
	Args []string
}

// Player opens a stream URL in a video player.
type Player interface {
	Open(url string) error
}

var genericPlayers = []GenericPlayer{
	{Name: "VLC", Args: []string{"vlc"}},
	{Name: "MPV", Args: []string{"mpv"}},
	{Name: "MPlayer", Args: []string{"mplayer"}},
}

// Open the given stream in a GenericPlayer.
func (p GenericPlayer) Open(url string) error {
	command := []string{}
	if runtime.GOOS == "darwin" {
		command = []string{"open", "-a"}
	}
	if runtime.GOOS == "windows" {
		command = []string{"cmd", "/C", "start"}
	}
	command = append(command, p.Args...)
	command = append(command, url)
	// #nosec
	// It is the user's responsibility to pass the correct arguments to open the url.
	return exec.Command(command[0], command[1:]...).Start()
}

// openPlayer opens a stream using the specified player and port.
func openPlayer(playerName string, port int) {
	var player Player
	for _, genericPlayer := range genericPlayers {
		if strings.EqualFold(genericPlayer.Name, playerName) {
			player = genericPlayer
			break
		}
	}
	if player == nil {
		log.Printf("Player '%s' is not supported. Currently supported players are: %s", playerName, joinPlayerNames())
		return
	}
	log.Printf("Playing in %s", playerName)
	if err := player.Open("http://localhost:" + strconv.Itoa(port)); err != nil {
		log.Printf("Error opening %s: %s\n", playerName, err)
	}
}

// joinPlayerNames returns a list of supported video players ready for display.
func joinPlayerNames() string {
	names := make([]string, len(genericPlayers))
	for i := range genericPlayers {
		names[i] = genericPlayers[i].Name
	}
	return strings.Join(names, ", ")
}


src/fileEntry.go
package src

import (
	"io"

	"github.com/anacrolix/torrent"
)

// SeekableContent describes an io.ReadSeeker that can be closed as well.
type SeekableContent interface {
	io.ReadSeeker
	io.Closer
}

// FileEntry helps reading a torrent file.
type FileEntry struct {
	*torrent.File
	torrent.Reader
}

// Seek seeks to the correct file position, paying attention to the offset.
func (f *FileEntry) Seek(offset int64, whence int) (int64, error) {
	return f.Reader.Seek(offset+f.File.Offset(), whence)
}

// NewFileReader sets up a torrent file for streaming reading.
func NewFileReader(f *torrent.File) (SeekableContent, error) {
	torrent := f.Torrent()
	reader := torrent.NewReader()

	// We read ahead 1% of the file continuously.
	reader.SetReadahead(f.Length() / 100)
	reader.SetResponsive()
	_, err := reader.Seek(f.Offset(), io.SeekStart)

	return &FileEntry{
		File:   f,
		Reader: reader,
	}, err
}


src/peerwatch.go
package src

import (
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"net"
	"net/http"
	"os"
	"os/exec"
	"os/signal"
	"regexp"
	"runtime"
	"strconv"
	"strings"
	"time"

	"github.com/dustin/go-humanize"
	"github.com/webview/webview"
)

const (
	VERSION = "v2.2"
	RESET   = "\033[0m"
	RED     = "\033[31m"
	GREEN   = "\033[32m"
	CYAN    = "\033[36m"
	PURPLE  = "\033[35m"
	BOLD    = "\033[1m"
	BLINK   = "\033[5m"
)

var (
	port         = flag.Int("p", 8080, "Port on which the tool will stream")
	tport        = flag.Int("tp", 50010, "Port on which the engine will work")
	tcp          = flag.Bool("tcp", true, "Connection over TCP or UDP")
	player       = flag.String("player", "vlc", "Video player to play video [vlc | mpv | mplayer]")
	link_regex   = "torrent/[0-9]{7}/[a-zA-Z0-9?%-]*/"
	magnet_regex = "magnet:\\?xt=urn:btih:[a-zA-Z0-9]*"
	query        string
	client       Client
	w            webview.WebView
	err          error
	ui           bool
)

var ch = make(chan string)
var UIclose = make(chan bool)

func Peerwatch(ln net.Listener) {
	windw()
	interrupt()
	banner()
	checkUpdates()
	flag.Usage = func() {
		fmt.Print(GREEN + "\n----------------------------------" + PURPLE + " PeerWatch " + GREEN + "----------------------------------\n" + RESET)
		fmt.Print(BOLD + CYAN + "[+] ./peerwatch " + RED + "[name]" + CYAN + " or ./peerwatch \"" + RED + "[name with multiple words]" + CYAN + "\"\n" + RESET)
		fmt.Print(BOLD + CYAN + "[+] Press " + RED + "ctrl + c" + CYAN + " to close\n\n" + RESET)
		flag.PrintDefaults()
		fmt.Print(GREEN + "----------------------------------" + PURPLE + " PeerWatch " + GREEN + "----------------------------------\n\n" + RESET)
	}
	flag.Parse()
	if len(flag.Args()) == 0 {
		ui = true
		flag.Usage()
		// Run GUI as a thread
		go gui(ln)
		query = <-ch
	} else {
		query = flag.Arg(0)
	}
	query = strings.ReplaceAll(strings.TrimSpace(query), " ", "+")
	name, magnet := find(query)
	// Found or Not
	if magnet == "" {
		fmt.Println(BOLD + RED + "[-] " + BLINK + "File Not Found" + RESET + RED + "...Try with another name" + RESET)
		if ui {
			w.Eval(`document.getElementById('status').innerText='File Not Found!!'`)
		}
		time.Sleep(time.Second)
		os.Exit(0)
	}
	if ui {
		w.Eval(fmt.Sprintf("document.getElementById('form').innerText='%s'", strings.Split(name, "/")[2]))
	}
	// Engine config
	cfg := ClientConfig{
		TorrentPath:    magnet,
		Port:           *port,
		TorrentPort:    *tport,
		Seed:           false,
		TCP:            *tcp,
		MaxConnections: 200,
	}
	engine(cfg)
}

func callback(uiname string) {
	ch <- uiname
}

func gui(ln net.Listener) {
	debug := false
	w = webview.New(debug)
	defer w.Destroy()
	w.SetTitle("PeerWatch")
	w.SetSize(480, 320, webview.HintNone)
	w.Bind("callback", callback)
	w.Navigate(fmt.Sprintf("http://%s/www/", ln.Addr()))
	w.Run()
	UIclose <- true
}

func find(query string) (string, string) {
	var magnet, name string
	dom := []string{"wtf", "st", "unblockninja.com", "unblockit.app", "eu", "ws", "se", "is", "gd", "tw", "buzz"}
	for i, d := range dom {
		log.Println("contacting server", i+1)
		// Get searches from query
		url := fmt.Sprintf("https://1337x.%s/search/%v/1/", d, query)
		resp, err := http.Get(url)
		if err != nil {
			log.Println("unable to connect to server", i+1)
			continue
		}
		defer resp.Body.Close()
		if resp.StatusCode == 200 {
			body, err := ioutil.ReadAll(resp.Body)
			if err != nil {
				log.Fatalf("unable to read defined search")
			}
			// regex find link
			re := regexp.MustCompile(link_regex)
			name = re.FindString(string(body))

			// Get magnet links from top search result
			get_magnet := fmt.Sprintf("https://1337x.%s/%v", d, name)
			get_magnet_resp, err := http.Get(get_magnet)
			if err != nil {
				log.Fatalf("unable to connect to domain")
			}
			defer get_magnet_resp.Body.Close()
			// if get_magnet_resp.StatusCode == 200 {
			magnet_body, err := ioutil.ReadAll(get_magnet_resp.Body)
			if err != nil {
				log.Fatalf("unable to read defined result")
			}
			// regex find magnet
			re = regexp.MustCompile(magnet_regex)
			magnet = re.FindString(string(magnet_body))
			// }
			return name, magnet
		}
	}
	return "", ""
}

func engine(cfg ClientConfig) {
	// Start torrent client
	client, err = NewClient(cfg)
	if err != nil {
		log.Fatalf(err.Error())
		os.Exit(0)
	}
	// Http handler
	go func() {
		http.HandleFunc("/", client.GetFile)
		log.Fatal(http.ListenAndServe(":"+strconv.Itoa(cfg.Port), nil))
	}()
	// Open VLC player
	go func() {
		for !client.ReadyForPlayback() {
			time.Sleep(time.Second)
		}
		openPlayer(*player, cfg.Port)
	}()
	counter := 0
	for {
		client.Render()
		if ui {
			counter = UIrender(counter)
		}
		time.Sleep(time.Second)
	}
}

func UIrender(counter int) int {
	data := humanize.Bytes(uint64(client.Progress))
	speed := DS
	perc := client.percentage()
	if client.Progress > 0 {
		w.Eval(fmt.Sprintf(`
		var TABLE = "<table>\
		  <tr>\
	        <td>Data Downloaded</td>\
		    <td>%v</td>\
		  </tr>\
		  <tr>\
		    <td>Speed</td>\
		    <td>%v</td>\
		  </tr>\
		  <tr>\
		    <td>Progress</td>\
		    <td>%.2f%%</td>\
		  </tr>\
		</table>";
		document.getElementById("progress").innerHTML = TABLE;
		`, data, speed, perc))

		w.Eval(fmt.Sprintf(`
		document.getElementById('status').innerHTML =
		'[Download the video once 100%% complete, at]<br><b> http://localhost:%v </b>';
		`, *port))
	} else {
		counter++
		w.Eval(fmt.Sprintf(`
		document.getElementById('status').innerHTML = 
		'Waiting to start: %v s<br>[It depends on your internet speed]'
		`, counter))
	}
	return counter
}

func banner() {
	bb := BOLD + GREEN + `
8888888b.                            888       888          888             888      
888   Y88b                           888   o   888          888             888      
888    888                           888  d8b  888          888             888      
888   d88P  .d88b.   .d88b.  888d888 888 d888b 888  8888b.  888888  .d8888b 88888b.    ` + PURPLE + `
8888888P"  d8P  Y8b d8P  Y8b 888P"   888d88888b888     "88b 888    d88P"    888 "88b 
888        88888888 88888888 888     88888P Y88888 .d888888 888    888      888  888   ` + GREEN + ` 
888        Y8b.     Y8b.     888     8888P   Y8888 888  888 Y88b.  Y88b.    888  888 
888         "Y8888   "Y8888  888     888P     Y888 "Y888888  "Y888  "Y8888P 888  888   ` + RED + `
====================================================================================   ` + CYAN + `
Created by:				                       Version: -~{ ` + RESET + RED + VERSION + BOLD + CYAN + ` }~-` + GREEN + `			    
+-+ +-+ +-+ +-+ +-+ +-+ +-+ +-+	
|C| |i| |p| |h| |e| |r| |a| |s|
+-+ +-+ +-+ +-+ +-+ +-+ +-+ +-+	                                                                                
` + RESET
	fmt.Print(bb + "\n")
}

func checkUpdates() {
	resp, err := http.Get("https://api.github.com/repos/cipheras/peerwatch/releases")
	if err != nil {
		log.Fatalln("unable to check updates:", err)
	}
	defer resp.Body.Close()
	data, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		log.Fatalln("unable to check updates")
	}
	var jsondata []map[string]interface{}
	err = json.Unmarshal(data, &jsondata)
	if err != nil {
		log.Fatalln("error parsing update data")
	}
	version := jsondata[0]["tag_name"].(string)
	releasName := jsondata[0]["name"].(string)
	if version != VERSION {
		fmt.Println(BOLD + CYAN + "[+] Update available..." + BLINK + GREEN + version + " [" + releasName + "]" + RESET)
	}
}

func windw() {
	if runtime.GOOS == "windows" {
		cmd := exec.Command("cmd")
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
		if err := cmd.Run(); err != nil {
			log.Println("windows color not working\n", err)
		}
		fmt.Printf("\033[2J\033[H")
	}
}

func interrupt() {
	c := make(chan os.Signal, 1)
	signal.Notify(c,
		os.Interrupt,
	)
	go func(c chan os.Signal) {
		select {
		case <-UIclose:
			log.Println("UI window closed")
			w.Terminate()
		case <-c:
		}
		fmt.Print("\r")
		log.Println("Exiting...!!")
		time.Sleep(time.Second)
		os.Exit(0)
	}(c)
}
